sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+        host: proxySettings.host,
+        port: proxySettings.port,
+        headers: request.headers.toJSON()
+      };
+      if (proxySettings.username && proxySettings.password) {
+        proxyAgentOptions.auth = `${proxySettings.username}:${proxySettings.password}`;
+      }
+      proxyAgent = new httpsProxyAgent(proxyAgentOptions);
+    }
+    return proxyAgent;
+  } else if (request.keepAlive) {
+    if (!keepAliveAgent) {
+      keepAliveAgent = new https.Agent({
+        keepAlive: true
+      });
+    }
+
+    return keepAliveAgent;
+  } else {
+    return https.globalAgent;
+  }
+}
+
+function isReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+class ReportTransform extends Transform {
+  private loadedBytes = 0;
+  private progressCallback: (progress: TransferProgressEvent) => void;
+  _transform(chunk: string | Buffer, _encoding: string, callback: Function): void {
+    this.push(chunk);
+    this.loadedBytes += chunk.length;
+    this.progressCallback({ loadedBytes: this.loadedBytes });
+    callback();
+  }
+
+  constructor(progressCallback: (progress: TransferProgressEvent) => void) {
+    super();
+    this.progressCallback = progressCallback;
+  }
+}
+
+/**
+ * A HttpsClient implementation that uses Node's "https" module to send HTTPS requests.
+ */
+export class NodeHttpsClient implements HttpsClient {
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const abortController = new AbortController();
+    let abortListener: ((event: any) => void) | undefined;
+    if (request.abortSignal) {
+      if (request.abortSignal.aborted) {
+        throw new AbortError("The operation was aborted.");
+      }
+
+      abortListener = (event: Event) => {
+        if (event.type === "abort") {
+          abortController.abort();
+        }
+      };
+      request.abortSignal.addEventListener("abort", abortListener);
+    }
+
+    if (request.timeout > 0) {
+      setTimeout(() => {
+        abortController.abort();
+      }, request.timeout);
+    }
+
+    if (request.formData) {
+      prepareFormData(request.formData, request);
+    }
+
+    if (!request.skipDecompressResponse) {
+      request.headers.set("Accept-Encoding", "gzip,deflate");
+    }
+
+    let body = request.body;
+
+    if (body && request.onUploadProgress) {
+      const onUploadProgress = request.onUploadProgress;
+      const uploadReportStream = new ReportTransform(onUploadProgress);
+      if (isReadableStream(body)) {
+        body.pipe(uploadReportStream);
+      } else {
+        uploadReportStream.end(body);
+      }
+
+      body = uploadReportStream;
+    }
+
+    if (body) {
+      if (isReadableStream(body)) {
+        request.headers.set("Transfer-Encoding", "chunked");
```

Comment:
Hmm you're right that I maybe have this part implemented incorrectly (I may have to write the size of each chunk to the stream, for example.)

The reason I ended up going this direction was because node-fetch will always use chunked encoding when it encounters a stream on req.body:

https://www.npmjs.com/package/node-fetch#default-headers

Reply:
Also the docs for the `http` module say

> Sends a chunk of the body. By calling this method many times, a request body can be sent to a server. In that case, it is suggested to use the ['Transfer-Encoding', 'chunked'] header line when creating the request.

But what they didn't say (but StackOverflow says) is that Node will automatically set Transfer-Encoding to chunked for you, as long as you don't set a Content-Length.

Reply:
so I think we can remove this

Reply:
Ah, I actually forgot that node does that automatically. Thanks for digging into that!

Reply:
I also did some work to set content-length more intelligently (which you'll see in a moment)